[nix] searxng recipe for installation via nix

efim opened the pull request:
This PR adds recipe for installing searxng via nix package manager, so that admins that use searx service on their nix managed servers could easily change it to searxng


## What does this PR do?

Adds:
- searxng package : "searxng" python description that's created with [mach-nix](https://github.com/DavHau/mach-nix), i.e extracts dependencies from `requirements.txt` of the project
- overlay for easier installation into nix configuration
- development shell with make & python with dependencies for local execution of built artifact

<!-- explain the changes in your PR, algorithms, design, architecture -->

## Why is this change important?
Searx is packaged for Nix, and can be installed & configured easily, [here sample](https://github.com/efim/nix-searx-container-example/blob/master/flake.nix)
This change would allow admins to easily use SearXNG with minimal modification of the configurations on their side.

<!-- MANDATORY -->

<!-- explain the motivation behind your PR -->

## How to test this PR locally?

<!-- commands to run the tests or instructions to test the changes-->

If Nix is installed (available for Linux & MacOS), then
- modify `./searx/settings.yml` secret value to non-default
- `nix build` would build the project, and symlink it to `./result`
- `nix-shell` would create subshell with env var SEARX_SETTINGS_PATH pointing to `./searx/settings.yml` and add `searx-run` to PATH
- `searx-run` would launch built project


## Author's checklist

<!-- additional notes for reviewiers -->
Using `mach-nix` to create recipe for SearXNG is different from the way Searx is packaged currently.
Mach nix takes exact versions of python dependencies from the `requirements.txt`, while [Searx definition](https://github.com/NixOS/nixpkgs/blob/nixos-21.05/pkgs/servers/web-apps/searx/default.nix#L71) takes versions of python libraries that are available in "nix store" and not necessarily match requirements (as I understand)

Also using mach-nix allows for easier maintaining of the nix recipe.
- `nix flake lock --update-input nixpkgs ` would pin "nixpkgs" to latest version
- any change to "requirements.txt" would automatically get picked up by the `nix build`

Downside: I spent quite a while trying to figure out how to add plugins to SearXNG that is packaged this way and wasn't successful, yet hopeful that will be able to figure it out in the future.
For now [nix packaged external Searx plugin](https://github.com/return42/tgwf-searx-plugins/pull/1) would only work with Searx

I suspect that writing more extensive documentation on SearXNG installation with nix would be useful,
but I created this PR with code first, to get the feedback and comments

## Related issues

<!--
Closes #234
-->

efim committed:
[nix] searxng recipe for installation via nix

- flake & compatibility
- searxng package & overlay for installation via Nix package manager
- development shell with make & python with dependencies for local
  execution of built artifact

dalf commented on flake.nix:
```diff
@@ -0,0 +1,99 @@
+                gnumake
+              ];
+              settingsFile = builtins.readFile ./searx/settings.yml;
+              SEARX_SETTINGS_PATH = "${settingsFile}";
```
`SEARX_SETTINGS_PATH` works, but the new value is `SEARXNG_SETTINGS_PATH`

efim replied:
Hello!

This line `settingsFile = builtins.readFile ./searx/settings.yml;` was intended by me for usage in development shell, 
so that locally build SearXNG could be launched easily after the repository checkout with only
1. Set new secret in ./searx/config.yml
2.
 ```
$ nix build
$ nix-shell
$ ./searx-run
```

And if a developer already has `SEARXNG_SETTINGS_PATH` or `SEARX_SETTINGS_PATH` set in their environment, then they can do
```
$ nix buld
# ./result/bin/searx-run
```
Since right now only reason `searx-run` is not successful is error because of default secret

It was not my intention to create a package of SearXNG with preattached configuration.

Usually admins that would install Searx with Nix would do so by providing configuration in the nix file, by setting some of the provided options ( [link with option definitions](https://github.com/NixOS/nixpkgs/blob/nixos-21.05/nixos/modules/services/networking/searx.nix) ,[ link with example configuration that installs searx & configures it](https://github.com/efim/nix-searx-container-example/blob/master/non-flake-config.nix#L49) )

And there is also some kind of configuration for optionally using uswgi in Searx module for NixOS, I suspect that it is being maintained.
As I understand (and tested a little bit) reusing same Searx service configuration with "pkg" modified to SearXNG seems to work well right now

dalf reviewed (commented):
> modify ./searx/settings.yml secret value to non-default

SearXNG looks for `settings.yml` in:
* `SEARXNG_SETTINGS_PATH` environment variable.
* `/etc/searxng/settings.yml` (actually not, but it should).
* `/etc/searx/settings.yml`
* `settings.yml` next to `webapp.py`

Perhaps, before the last step, it should try on `~/.config/searxng/settings.yml`?

> Mach nix takes exact versions of python dependencies from the requirements.txt

This is the thing to do: SearXNG is tightly couple to `httpx[http2]` and `httpx-socks[asyncio]`

> Downside: I spent quite a while trying to figure out how to add plugins to SearXNG that is packaged this way

What are the difficulties? May be we can help.
If the copy of the static files is a problem:
* define a deployement directory
* set settings.yml

The thing that would be interresting perhaps in a second step if how to deploy the application with uwsgi.

efim committed:
[nix] explicit pypi-deps-db flake input

This change will allow for easier fix of the situation when
python dependencies of the project get updated to version that's not
available in current version of "pypi-deps-db", and also not yet
propagated to "mach-nix-src" repo

This way "nix flake lock --update-input pypi-deps-db" will take latest
commit of the db, and it will be used as input for "mach-nix-src" locally
